[CBRD-26246] CREATE TABLE 시 REPLICATION 옵션 추가

youngjun9072 opened the pull request:
[<http://jira.cubrid.org/browse/CBRD-26246>](http://jira.cubrid.org/browse/CBRD-26246)

### Purpose

PK 제약사항 변경을위한 REPLICATION 옵션 문법을 추가합니다.


### Implementation

테이블 생성 시 다음 옵션 및 값을 허용합니다.
- CREATE TABLE tbl();
- CREATE TABLE tbl() REPLICATION;
- CREATE TABLE tbl() REPLICATION = ON | OFF;
- CREATE TABLE tbl() REPLICATION ON | OFF; 
- 옵션을 입력하지 않을 경우 기본 값은 ON 으로 처리합니다.

테이블 수정 시 다음 옵션 및 값을 허용 합니다.
- ALTER TABLE tbl REPLICATION;
- ALTER TABLE tbl REPLICATION=ON | OFF;
- ALTER TABLE tbl REPLICATION ON OFF;
CREATE 와 다르게 REPLICATION 옵션을 입력하지 않을 경우 아무 처리도 하지 않습니다

주요 구현은 다음과 같습니다.
"replication" 키워트 입력 추가 
- csql_lexer.l 파일에 replication 키워드를 대소문자 구분없이 인식할 수 있도록 추가

PT_TABLE_OPTION_REPLICATION 옵션 타입 추가
- parse_tree.h 에 PT_TABLE_OPTION_REPLICATION 옵션 타입 추가

REPLICATION 옵션이 반복적으로 등장하는지 확인
- semantic_check.c 파일의 pt_check_create_entity 함수에서 REPLICATION 키워드가 반복적으로 명시됐는지 확인

REPLICATION 옵션 OFF 확인 및 로깅
- 하위버전에서 백업 데이터를 최신 큐브리드로 복원 시 데이터는 REPLICATION 정보가 존재하지 않으므로 모두 복제 대상이 아닌것으로 처리됨
- 기본으로 복제 대상으로 처리하기위해 NO_REPLICATION 플래그를 만들어 복제 대상이 아닌 테이블만 별도로 표시
- 하위버전에서 데이터를 복원했을때 NO_REPLICATION 플래그가 설정되어있지 않으므로 모두 복제 대상으로 복구됨

### Remarks

이 이슈는 문법 추가만 다루는 이슈이고 핸들링은 다음 이슈에서 처리합니다.

youngjun9072 committed:
add replication option

youngjun9072 committed:
lint

youngjun9072 committed:
fix typo

youngjun9072 committed:
Add variable intialization

youngjun9072 committed:
add "alter table ... replication" option

youngjun9072 committed:
lint

youngjun9072 committed:
add "TODO" comment

InChiJun commented on src/parser/parse_tree.h:
```diff
@@ -2056,6 +2058,10 @@ struct pt_alter_info
     } comment;
+    struct
+    {
+      PT_NODE *tbl_replication;
```
사전에 말씀드린대로, 2063라인에 주석을 추가하면 더 좋지 않을까 싶습니다.

youngjun9072 replied:
반영됐습니다.

youngjun9072 committed:
Code changes

- Changed the location where replication option is handled
- Simplified condition logic to check only default or "on"
- Added explanatory comments

youngjun9072 committed:
lint

hornetmj commented:
@youngjun9072 'CREATE 와 다르게 REPLICATION 옵션을 입력하지 않을 경우 아무 처리도 하지 않습니다' 관련해서 사용자는 성공했다고 생각해야 할까요? 아니면 실패했다고 생각해야 할까요? 정확한 하나의 문법만 허용하는게 어떨까요?

hornetmj commented on src/parser/csql_lexer.l:
```diff
@@ -881,6 +881,9 @@ IDL       [a-zA-Z0-9_]
 [dD][iI][sS][kK][_][sS][iI][zZ][Ee]				{ begin_token(yytext);
 										csql_yylval.cptr = pt_makename(yytext);
 										return DISK_SIZE; }
+[rR][eE][pP][lL][iI][cC][aA][tT][iI][oO][nN]				{ begin_token(yytext);
```
정렬해주세요.

youngjun9072 replied:
이 파일의 정렬 기준은 어느것을 참고하면될까요?

hornetmj replied:
렉서 파일은 도구를 사용하여 정렬하지 않기 때문에 위 아래와 상대적인 정렬을 수행하시면 됩니다. 위 경우 tab에 의해 차이가 나는 것으로 보입니다. vi 기준 :set list 후 ^I 부분을 확인하시고 맞추시면 정렬될 것 같습니다.

youngjun9072 replied:
반영했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -66,6 +66,7 @@
 #define strlen(s1)  ((int) strlen(s1))
 #endif /* defined (SUPPRESS_STRLEN_WARNING) */
 
+
```
```suggestion
```

youngjun9072 replied:
반영했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -9214,6 +9225,19 @@ do_create_entity (PARSER_CONTEXT * parser, PT_NODE * node)
+       * If the option is omitted, or if the "on|off" value is omitted,  
+       * the default is set to "on".  
+       */
+      if (REPL_IS_ON_OR_DEFAULT (tbl_opt_replication))
```
NIT:
- 디폴트가 ON이기 때문에 구분할 필요가 없을 것 같습니다.
- 코드 가독성을 위해 다른 테이블 옵션과 동일한 또는 유사한 코드 형식으로 구현하는게 어떨까요?

youngjun9072 replied:
다른 테이블 옵션을 사용하는 경우는 이 코드에서 현재 두가지 방법입니다
첫번째는 tbl_opt_xxx를 확인하는 방법인데, 일반적인 옵션은 입력을 해야만 동작을하는 구조이기때문에 tbl_opt_xxx에 값이 존재함으로써 판단할 수 있지만 replication은 옵션 생략시 혹은 옵션에 값이 있을 경우에도 핸들링을 해야합니다.
두번째는 reuse_oid 과 같은 방식인데 오히려 이 방법이 reuse_oid 여부를 파라미터로 확인하기위해 사용되고 있어 다른 옵션의 패턴과 일치한다는 생각은 들지 않습니다. 

매크로로인해 가독성이 떨어진다면 다른 리뷰어들 의견 반영해 매크로를 지우던가 bool 로 처리하도록하겠습니다.

hornetmj replied:
전달드렸던 개발 가이드 참고해보시기 바랍니다.
- https://github.com/CUBRID/cubrid/commit/5739e20816cfbbf9e3d4e1be24b68b8b28ce8bfc#diff-05bddf8d49a888bf00c32b386b747c8fd1e79104383dd5642485ab380d63e4e1

youngjun9072 committed:
리뷰 반영
 - 매크로 이름 변경
 - 불필요 공백 삭제

youngjun9072 commented:
@hornetmj "alter table tbl;" 로 입력했을 경우 현재 코드에서는 아무 처리를 하지 않고 기존 문법에러가 뜸을 의미합니다.
하나의 문법만 허용한다고하심은 어떤 예시가 있을까요???

hornetmj commented:
> @hornetmj "alter table tbl;" 로 입력했을 경우 현재 코드에서는 아무 처리를 하지 않고 기존 문법에러가 뜸을 의미합니다. 하나의 문법만 허용한다고하심은 어떤 예시가 있을까요???

'CREATE 와 다르게 REPLICATION 옵션을 입력하지 않을 경우 아무 처리도 하지 않습니다'에서 '아무 처리도 하지 않습니다'라는 말이 기존과 동일하게 문법 에러가 발생한다는 의미라면 스펙에 동의합니다. 문법 관련해서는 현 구현이 REPLICATION, REPLICATION ON, REPLICATION=ON 등을 모두 허용하고 있어서 명시적으로 하나만 허용하자는 커멘트 였습니다. 그러나 ENCRYPT나 COMMENT 등 다른 테이블 옵션을 참고했을 때 그대로 두셔도 무방할 것 같습니다.

youngjun9072 committed:
add tab

youngjun9072 committed:
Flag and condition changes for backward compatibility with older backups

- replace SM_CLASSFLAG_REPLICATION with SM_CLASSFLAG_NO_REPLICATION.
- change handling condition from IS_REPLICATION_ON to !IS_REPLICATION.

youngjun9072 committed:
Add macro for alter condition branch
